File: 250321/안전 지대/comfort-zone.cpp
#include <bits/stdc++.h>
using namespace std;

#define MAX_NM 50
#define DIR_num 4

int n, m;
int grid[MAX_NM][MAX_NM];
bool visited[MAX_NM][MAX_NM];

int max_k, ans_k, max_save, save_c;

bool CanGo(int x, int y, int k) {
    return 0<=x && x<n && 0<=y && y<m && grid[x][y] > k && !visited[x][y];
}

void DFS(int x, int y, int k) {
    int dx[DIR_num] = {-1, 0, 1, 0};
    int dy[DIR_num] = {0, 1, 0, -1};

    for(int d=0; d<DIR_num; d++) {
        int nx = x+dx[d], ny = y+dy[d];

        if(CanGo(nx, ny, k)) {
            visited[nx][ny] = true;
            DFS(nx, ny, k);
        }
    }
}

int main() {
    ios::sync_with_stdio(0);
    cin.tie(0);

    cin >> n>>m;

    for(int i=0; i<n; i++) {
        for(int j=0; j<m; j++) {
            cin >> grid[i][j];

            max_k = max(max_k, grid[i][j]);
        }
    }

    for(int k=1; k<max_k; k++) {
        save_c = 0;
        for(int i=0; i<n; i++) {
            for(int j=0; j<m; j++) {
                visited[i][j] = false;
            }
        }

        for(int i=0; i<n; i++) {
            for(int j=0; j<m; j++) {
                if(grid[i][j] > k && !visited[i][j]) {
                    visited[i][j] = true;
                    DFS(i, j, k);
                    save_c++;
                }
            }
        }

        if(save_c > max_save) {
            max_save = save_c;
            ans_k = k;
        }
    }

    cout << ans_k << ' ' << max_save;
    return 0;
}

File: 250404/안전 지대/comfort-zone.py
n, m = tuple(map(int, input().split()))

grid = [
    list(map(int, input().split()))
    for _ in range(n)
]

visited = [[False]*m for _ in range(n)]

max_k = max([
    grid[i][j]
    for i in range(n)
    for j in range(m)
])

def can_go(x, y, k):
    return 0<=x<n and 0<=y<m and grid[x][y] > k and not visited[x][y]

def dfs(x, y, k):
    dxs, dys = [-1, 0, 1, 0], [0, 1, 0, -1]

    for dx, dy in zip(dxs, dys):
        nx, ny = x+dx, y+dy

        if can_go(nx, ny, k):
            visited[nx][ny] = True
            dfs(nx, ny, k)

ans_k = 0
ans = 0


for k in range(1, max_k+1):
    visited = [[False]*m for _ in range(n)]
    cnt = 0

    for i in range(n):
        for j in range(m):
            if can_go(i, j, k):
                visited[i][j] = True
                dfs(i, j, k)

                cnt += 1
    
    if cnt > ans:
        ans_k = k
        ans = cnt

print(ans_k, ans)